File: abstract-factory/abstract-factory.go
package main

import "fmt"

// Concreate Elements implementation
type Car interface {
	getPrice() int
	getMaxLoadWeigth() int
	calculateOutcome() int
}

type BMWCar struct {
}

func (c *BMWCar) getPrice() int {
	return 0
}
func (c *BMWCar) getMaxLoadWeigth() int {
	return 0
}
func (c *BMWCar) calculateOutcome() int {
	return 0
}

type MercCar struct {
}

func (c *MercCar) getPrice() int {
	return 0
}
func (c *MercCar) getMaxLoadWeigth() int {
	return 0
}
func (c *MercCar) calculateOutcome() int {
	return 0
}

type Boat interface {
	getPrice() int
	getMaxLoadWeigth() int
	calculateOutcome() int
	getSizeOfBoat() []float64
}

type BMWBoat struct {
}

func (c *BMWBoat) getPrice() int {
	return 0
}
func (c *BMWBoat) getMaxLoadWeigth() int {
	return 0
}
func (c *BMWBoat) calculateOutcome() int {
	return 0
}
func (c *BMWBoat) getSizeOfBoat() []float64 {
	return []float64{}
}

type MercBoat struct {
}

func (c *MercBoat) getPrice() int {
	return 0
}
func (c *MercBoat) getMaxLoadWeigth() int {
	return 0
}
func (c *MercBoat) calculateOutcome() int {
	return 0
}
func (c *MercBoat) getSizeOfBoat() []float64 {
	return []float64{}
}

// / abstraction implementation
type TransportFactory interface {
	createCar() Car
	createBoat() Boat
}

type MercFactory struct {
}

func (f *MercFactory) createCar() Car {
	return &MercCar{}
}
func (f *MercFactory) createBoat() Boat {
	return &MercBoat{}
}

type BMWFactory struct {
}

func (f *BMWFactory) createCar() Car {
	return &BMWCar{}
}
func (f *BMWFactory) createBoat() Boat {
	return &BMWBoat{}
}

// / client implementation
type Transports struct {
	car  Car
	boat Boat
}

func CreateTransport(f TransportFactory) *Transports {
	return &Transports{
		car:  f.createCar(),
		boat: f.createBoat(),
	}
}

func main() {
	t := CreateTransport(&BMWFactory{})

	totalOutcome := t.car.calculateOutcome() + t.boat.calculateOutcome()

	fmt.Println(totalOutcome)
}

File: abstract-factory/abstract-factory.php
<?php

// Concreate Elements implementation
interface Car
{
    public function getPrice(): int;
    public function getMaxLoadWeigth(): int;
    public function calculateOutcome(): int;
}

class BMWCar implements Car
{
    public function getPrice(): int
    {
        return 0;
    }
    public function getMaxLoadWeigth(): int
    {
        return 0;
    }
    public function calculateOutcome(): int
    {
        return 0;
    }
}

class MercCar implements Car
{
    public function getPrice(): int
    {
        return 1;
    }
    public function getMaxLoadWeigth(): int
    {
        return 1;
    }
    public function calculateOutcome(): int
    {
        return 1;
    }
}


interface Boat
{
    public function getPrice(): int;
    public function getMaxLoadWeigth(): int;
    public function calculateOutcome(): int;
    public function getSizeOfBoat(): array;
}

class BMWBoat implements Boat
{
    public function getPrice(): int
    {
        return 0;
    }
    public function getMaxLoadWeigth(): int
    {
        return 0;
    }
    public function calculateOutcome(): int
    {
        return 0;
    }
    public function getSizeOfBoat(): array
    {
        return  [];
    }
}

class MercBoat implements Boat
{
    public function getPrice(): int
    {
        return 1;
    }
    public function getMaxLoadWeigth(): int
    {
        return 1;
    }
    public function calculateOutcome(): int
    {
        return 1;
    }
    public function getSizeOfBoat(): array
    {
        return  [];
    }
}


/// abstraction implementation

interface TransportFactory
{
    public function createCar(): Car;
    public function createBoat(): Boat;
}

class BMWFactory implements TransportFactory
{

    public function createCar(): Car
    {
        return new BMWCar;
    }

    public function createBoat(): Boat
    {
        return new BMWBoat;
    }
}

class MercFactory implements TransportFactory
{

    public function createCar(): Car
    {
        return new MercCar;
    }

    public function createBoat(): Boat
    {
        return new MercBoat;
    }
}

/// Client implementation

class ClientService
{

    protected Car $car;
    protected Boat $boat;

    public function __construct(TransportFactory $factory)
    {
        $this->car = $factory->createCar();
        $this->boat = $factory->createBoat();
    }

    public function calculateAllOutcome(): int
    {
        return $this->car->calculateOutcome() + $this->boat->calculateOutcome();
    }
}

class ClientConstructor
{
    protected ClientService $service;

    // And also you can inject ClientService there
    public function __construct()
    {
        $this->service = new ClientService(new MercFactory);
    }

    public function printAllOutcome(): void
    {
        print($this->service->calculateAllOutcome());
    }
}
